composition/src/index.js

```diff
@@ -0,0 +1,260 @@
+import { options } from 'preact';
+
+const EMPTY_STATE = {};
+
+/** @type {import('./internal').Component} */
+let currentComponent;
+
+let oldBeforeRender = options._render;
+options._render = vnode => {
+	if (oldBeforeRender) oldBeforeRender(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = (currentComponent = vnode._component);
+
+	/** the vnode component is a uninitialized composition component */
+	if (c.constructor.__compositions && !c.__compositions) {
+		c.__compositions = { u: [], w: [], e: [], x: {} };
+		const render = c.constructor(() => c.props);
+		c.constructor = props => render(props, c.__compositions.r);
+	}
+
+	/** the vnode component is a composition initialized component */
+	if (c.__compositions) {
+		// Update the ref to forward
+		if (vnode.ref) {
+			c.__compositions.r = vnode.ref;
+			vnode.ref = null;
+		}
+
+		// call all watch
+		c.__compositions.w.forEach(up => handleEffect(up, c));
+	}
+};
+
+let oldAfterDiff = options.diffed;
+options.diffed = vnode => {
+	if (oldAfterDiff) oldAfterDiff(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = vnode._component;
+	if (c && c.__compositions)
+		// handle all `effect`s
+		c.__compositions.e.forEach(up => handleEffect(up, c));
+};
+
+let oldBeforeUnmount = options.unmount;
+options.unmount = vnode => {
+	if (oldBeforeUnmount) oldBeforeUnmount(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = vnode._component;
+	if (c && c.__compositions) {
+		// cleanup `effect`s onCleanup
+		c.__compositions.e.forEach(cleanupEffect);
+		// call all onUnmounted lifecycle callbacks
+		c.__compositions.u.forEach(f => f());
```

Comment:
Bundle size nits, if you aren't returning anything from `cleanupEffect` and `f` it's better to do { f() } since now your transpilation will add `return` in front of it increasing bundle size.
Also `some` will save you some extra bytes compared to `forEach`

Reply:
That removed 6b from bundle, thanks